lib/SIL/IR/SILFunctionType.cpp

```diff
@@ -1345,20 +1363,34 @@ class Conventions {
     llvm_unreachable("unhandled ownership");
   }
 
-  // Determines owned/unowned ResultConvention of the returned value based on
-  // returns_retained/returns_unretained attribute.
+  // Determines the ownership ResultConvention (owned/unowned) of the return
+  // value using the SWIFT_RETURNS_(UN)RETAINED annotation on the C++ API; if
+  // not explicitly annotated, falls back to the
+  // SWIFT_RETURNS_(UN)RETAINED_BY_DEFAULT annotation on the C++
+  // SWIFT_SHARED_REFERENCE type.
   std::optional<ResultConvention>
   getCxxRefConventionWithAttrs(const TypeLowering &tl,
                                const clang::Decl *decl) const {
-    if (tl.getLoweredType().isForeignReferenceType() && decl->hasAttrs()) {
-      for (const auto *attr : decl->getAttrs()) {
-        if (const auto *swiftAttr = dyn_cast<clang::SwiftAttrAttr>(attr)) {
-          if (swiftAttr->getAttribute() == "returns_unretained") {
-            return ResultConvention::Unowned;
-          } else if (swiftAttr->getAttribute() == "returns_retained") {
-            return ResultConvention::Owned;
-          }
-        }
+    if (!tl.getLoweredType().isForeignReferenceType())
+      return std::nullopt;
+
+    if (auto result = matchSwiftAttr<ResultConvention>(
+            decl, {{"returns_unretained", ResultConvention::Unowned},
+                   {"returns_retained", ResultConvention::Owned}})) {
+      return result;
+    }
+
+    if (auto *classDecl = tl.getLoweredType()
+                              .getASTType()
```

Comment:
`getASTType` has a huge number of of caveats in the documentation. Would it be possible to get the return type from the `clang::Decl` instead?